### ISSUE TITLE
React Native SDK: application stops responding while using flexible-sync adding subscription WHEN a record is deleted from mongoDB atlas.

### ISSUE THREAD

**Issue by akashatnix** [realm/realm-js#4768]
### How frequently does the bug occur?

All the time

### Description

**I am using React native SDK:**

### My Setup Overview:
<br/>

_package.json_
```
"@realm/react": "^0.3.2",
"realm": "^10.19.5",
```

all my components are wrapped under `AppProvider` and `RealmProvider` as follows
```
const syncConfig = {
    flexible: true,
    error: (session, error) => console.log("error: ", error),
    newRealmFileBehavior: {type: "openImmediately"},
    existingRealmFileBehavior: {type: "openImmediately"}
};

<AppProvider id="...">
    <RealmProvider sync={syncConfig}>
        <MyComponent/>
    </RealmProvider>
</AppProvider>
```

`MyComponent` adds subscription as follows:
```
const realm = useRealm();
const results = useQuery("CollectionName");
useEffect(() => {
    realm.subscriptions.update(mutableSubs => mutableSubs.add("SOME_QUERY", {name: "NAME", throwOnUpdate: true}));
});
```
and then I render results in `<FlatList>` if length>0

#### Issue

Since the query is subscribed (as mentioned above). Now:

if i login into mongoDB Altas and delete a record from my collection, then my application stops responding. I am guessing that the react native realm SDK is trying to sync with the cloud blocking the main/ui thread of my application making it unresponsive. back button is unresponsive too. I had to kill the app. 

More than that, this issue persists even if i restart my app (can click other buttons). But, as soon as it adds the query to realm subscription, it becomes unresponsive.

There is no log on console once query is subscribed and app becomes unresponsive.

### Can you reproduce the bug?

Yes, always

### Version

"realm": "^10.19.5"

### What SDK flavour are you using?

Atlas Device Sync

### Are you using encryption?

No, not using encryption

### Platform OS and version(s)

Samsung Android S21+ 5G (SM-G986W - 12)

### Build environment

Which debugger for React Native: ..

**Comment by kneth** [comment]
@akashatnix Thank you for reporting.

I wonder if it makes a difference with you remove `newRealmFileBehavior` and `existingRealmFileBehavior` from the configuration.

If possible, it would be interesting if you observe it on iOS too.

**Comment by kneth** [comment]
You might want to increase the log level by calling `Realm.App.Sync.setLogLevel(app, level)` - see https://www.mongodb.com/docs/realm-sdks/js/latest/Realm.App.Sync.html. Moreover, on Android you will also have to choose a logger (`console.log()` can be used) - see `Realm.App.Sync.setLogger(app, logger)`.

```js
Realm.App.Sync.setLogLevel(app, "all");
Realm.App.Sync.setLogger(app, (level, message) => console.log(`[${level}] ${message}`));
```

**Comment by kneth** [comment]
By the way, do you have multiple `RealmProvider`s? (see https://github.com/realm/realm-js/issues/4715#issuecomment-1211911271)

**Comment by akashatnix** [comment]
@kneth - Nope. I do not have multiple `RealmProvider`s (Just one)

I enabled Realm App logger as follows
```
const app = new Realm.App({id: "<MY_APP_ID>", timeout: 10000});
Realm.App.Sync.setLogLevel(app, "all");
Realm.App.Sync.setLogger(app, (level, message) => console.log(`[${level}] ${message}`));
```

### When I had **not removed** `newRealmFileBehavior` and `existingRealmFileBehavior` - (as per logs: I assume Realm SDK wasn't able to resolve `54.145.94.221:443` as there were no further logs)
```
LOG  [2] User agent string: 'RealmSync/12.3.0 (Android Linux 4.19.113-23195093 #1 SMP PREEMPT Tue Jan 25 14:52:06 KST 2022 aarch64) RealmJS/10.19.5 (reactnative, android, v31) '
 LOG  [3] Connection[1]: Session[1]: Binding '/data/data/com.example.app/files/mongodb-realm/<MY_APP_ID>/62992b926ac6ddeb360d855d/flx_sync_default.realm' to ''
 LOG  [2] Connection[1]: Session[1]: Activating
 LOG  [4] Connection[1]: Session[1]: client_reset_config = false, Realm exists = true, client reset = false
 LOG  [2] Connection[1]: Session[1]: client_file_ident = 251, client_file_ident_salt = 3747123663050590884
 LOG  [1] Connection[1]: Session[1]: last_version_available  = 282
 LOG  [1] Connection[1]: Session[1]: progress_server_version = 4612
 LOG  [1] Connection[1]: Session[1]: progress_client_version = 5
 LOG  [2] Connection[1]: Session[1]: Progress handler called, downloaded = 1201, downloadable(total) = 1201, uploaded = 160, uploadable = 160, reliable_download_progress = false, snapshot version = 282
 LOG  [2] WebSocket::Websocket()
 LOG  [3] Resolving 'ws.us-east-1.aws.realm.mongodb.com:443'
 LOG  [2] Connection[1]: Session[1]: Progress handler called, downloaded = 1201, downloadable(total) = 1201, uploaded = 160, uploadable = 160, reliable_download_progress = false, snapshot version = 282
 LOG  [2] Connection[1]: Session[1]: Progress handler called, downloaded = 1201, downloadable(total) = 1201, uploaded = 160, uploadable = 160, reliable_download_progress = false, snapshot version = 282
 LOG  [3] Connecting to endpoint '54.145.94.221:443' (1/1)
```

### When I **removed** `newRealmFileBehavior` and `existingRealmFileBehavior`
```
LOG  [2] User agent string: 'RealmSync/12.3.0 (Android Linux 4.19.113-23195093 #1 SMP PREEMPT Tue Jan 25 14:52:06 KST 2022 aarch64) RealmJS/10.19.5 (reactnative, android, v31) '
 LOG  [3] Connection[1]: Session[1]: Binding '/data/data/com.example.app/files/mongodb-realm/<MY_APP_ID>/62992b926ac6ddeb360d855d/flx_sync_default.realm' to ''
 LOG  [2] Connection[1]: Session[1]: Activating
 LOG  [4] Connection[1]: Session[1]: client_reset_config = false, Realm exists = true, client reset = false
 LOG  [2] Connection[1]: Session[1]: client_file_ident = 251, client_file_ident_salt = 3747123663050590884
 LOG  [1] Connection[1]: Session[1]: last_version_available  = 282
 LOG  [1] Connection[1]: Session[1]: progress_server_version = 4612
 LOG  [1] Connection[1]: Session[1]: progress_client_version = 5
 LOG  [2] Connection[1]: Session[1]: Progress handler called, downloaded = 1201, downloadable(total) = 1201, uploaded = 160, uploadable = 160, reliable_download_progress = false, snapshot version = 282
 LOG  [2] WebSocket::Websocket()
 LOG  [3] Resolving 'ws.us-east-1.aws.realm.mongodb.com:443'
 LOG  [3] Connecting to endpoint '34.227.4.145:443' (1/1)
 LOG  [4] Connected to endpoint '34.227.4.145:443' (from '10.0.0.159:46248')
 LOG  [2] Verifying server SSL certificate using root certificates, host name = ws.us-east-1.aws.realm.mongodb.com, server port = 443, certificate =
-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----
 LOG  [4] Verifying server SSL certificate using 155 root certificates
 LOG  [2] Server SSL certificate verified using root certificate(37):
-----BEGIN CERTIFICATE-----
...
-----END CERTIFICATE-----
 LOG  [2] WebSocket::initiate_client_handshake()
 LOG  [1] HTTP request =
GET /api/client/v2.0/app/<MY_APP_ID>/realm-sync?baas_at=<MY_JWT> HTTP/1.1
Host: ws.us-east-1.aws.realm.mongodb.com
Connection: Upgrade
Sec-WebSocket-Key: mlkTwh17Zcuz22ccLIGwJg==
Sec-WebSocket-Protocol: com.mongodb.realm-query-sync/6, com.mongodb.realm-query-sync/5, com.mongodb.realm-query-sync/4, com.mongodb.realm-query-sync/3, com.mongodb.realm-query-sync/2
Sec-WebSocket-Version: 13
Upgrade: websocket
User-Agent: RealmSync/12.3.0 (Android Linux 4.19.113-23195093 #1 SMP PREEMPT Tue Jan 25 14:52:06 KST 2022 aarch64) RealmJS/10.19.5 (reactnative, android, v31)
 LOG  [2] WebSocket::handle_http_response_received()
 LOG  [1] HTTP response = HTTP/1.1 101 Switching Protocols
cache-control: no-cache, no-store, must-revalidate
connection: Upgrade
date: Thu, 11 Aug 2022 16:43:09 GMT
sec-websocket-accept: iZqDMNGHTQvmgRCdgDogFns8n5g=
sec-websocket-protocol: com.mongodb.realm-query-sync/6
server: envoy
upgrade: websocket
vary: Origin
x-frame-options: DENY
 LOG  [3] Connection[1]: Negotiated protocol version: 6
 LOG  [2] Connection[1]: Will emit a ping in 28568 milliseconds
 LOG  [2] Connection[1]: Session[1]: Sending: IDENT(client_file_ident=251, client_file_ident_salt=3747123663050590884, scan_server_version=4612, scan_client_version=5, latest_server_version=4612, latest_server_version_salt=1364306390683158842, query_version: 66 query_size: 123, query: "{"NAME":"SOME_QUERY"}")
 LOG  [2] Connection[1]: Session[1]: Sending: MARK(request_ident=2)
 LOG  [2] Connection[1]: Session[1]: Received: MARK(request_ident=2)
 LOG  [2] Connection[1]: Session[1]: Sending: UPLOAD(progress_client_version=282, progress_server_version=4612, locked_server_version=4612, num_changesets=0)
 LOG  [2] Connection[1]: Sending: PING(timestamp=31437428, rtt=0)
 LOG  [2] Connection[1]: Session[1]: Progress handler called, downloaded = 1201, downloadable(total) = 1201, uploaded = 160, uploadable = 160, reliable_download_progress = false, snapshot version = 282
 LOG  [2] Connection[1]: Session[1]: Progress handler called, downloaded = 1201, downloadable(total) = 1201, uploaded = 160, uploadable = 160, reliable_download_progress = false, snapshot version = 282
```
in the above logs as you can see `LOG  [2] Connection[1]: Sending: PING(timestamp=31437428, rtt=0)` - this does not appear every-time. Multiple times I killed and restarted my app, this log appears sometimes only.
When the milliseconds in ` LOG  [2] Connection[1]: Will emit a ping in 28568 milliseconds` is some higher number, then the `PING` log does not appear as my app had already became unresponsive.

**NOTE: In Both the cases (whether or not removed `newRealmFileBehavior` and `existingRealmFileBehavior`), my app became unresponsive**

Moreover This happens every-time now (not just when Record is deleted from mongoDB-atlas Collection). This happened even when I inserted new record in the mongoDB-atlas Collection.

**Comment by kneth** [comment]
While you are at it, adding a simple error handler to the sync configuration should be useful:

```js
sync: {
  error: (session, err) => console.log(`Sync error: ${JSON.stringify(err)}`),
  // ... more option
}
```

The logs shows that your app is downloading 1201 bytes, and thereafter nothing happens. 

> Moreover This happens every-time now (not just when Record is deleted from mongoDB-atlas Collection). This happened even when I inserted new record in the mongoDB-atlas Collection.

Do you have indications that the MongoDB translator paused (see also https://www.mongodb.com/docs/atlas/app-services/sync/error-handling/errors/#mongodb-translator-errors)? Not that it should make your app unresponsive.

**Comment by kneth** [comment]
You should contact [support](https://www.mongodb.com/products/enterprise-grade-support) as they also have access to the logs server side.

**Comment by kneth** [comment]
@akashatnix I am closing the issue, and I hope that our support has been able to provide a solution for you.

**Comment by akashatnix** [comment]
Hello,

Yes I got my issue resolved. Just for an update, I did not remove _newRealmFileBehavior_ and _existingRealmFileBehavior_ from the configuration. The real issue was that my subscription query was crashing as it was incorrectly formatted. The app was silently crashing without any logs. Later I had to put **await** while adding subscription and had to wrap it with **try-catch** as follows that helped me debug:

```
try {
  console.log(`adding query to subscribe '${collectionName}': `, query)
  await realm.subscriptions.update(mutableSubs => {
    mutableSubs.add(query, option);
  });
} catch (error) {
  console.log(`Error adding query subscription for collection "${collectionName}": `, error);
}
```

I think this documentation is missing in realm docs.
